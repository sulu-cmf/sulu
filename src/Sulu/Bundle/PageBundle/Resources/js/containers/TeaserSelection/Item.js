// @flow
import React, {Fragment} from 'react';
import {action, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import {observer} from 'mobx-react';
import textVersion from 'textversionjs';
import {MimeTypeIndicator} from 'sulu-media-bundle/components';
import {SingleMediaSelectionOverlay} from 'sulu-media-bundle/containers';
import type {Media} from 'sulu-media-bundle/types';
import {Button, Icon, Input} from 'sulu-admin-bundle/components';
import {TextEditor} from 'sulu-admin-bundle/containers';
import {translate} from 'sulu-admin-bundle/utils';
import teaserProviderRegistry from './registries/teaserProviderRegistry';
import itemStyles from './item.scss';
import type {TeaserItem} from './types';

type Props = {|
    description: ?string,
    edited: ?boolean,
    editing: boolean,
    id: number | string,
    locale: IObservableValue<string>,
    mediaId: ?number,
    onApply: (item: TeaserItem) => void,
    onCancel: (type: string, id: number | string) => void,
    title: ?string,
    type: string,
|};

@observer
class Item extends React.Component<Props> {
    static mediaUrl: ?string = undefined;

    @observable title: ?string = undefined;
    @observable description: ?string = undefined;
    @observable mediaId: ?number = undefined;
    @observable mediaOverlayOpen: boolean = false;

    componentDidMount() {
        this.setStateFromProps();
    }

    componentDidUpdate(prevProps: Props) {
        if (prevProps.title !== this.props.title
            || prevProps.description !== this.props.description
            || prevProps.mediaId !== this.props.mediaId
        ) {
            this.setStateFromProps();
        }

        if (prevProps.editing === true && this.props.editing === false) {
            this.setStateFromProps();
        }
    }

    @action setStateFromProps() {
        const {description, mediaId, title} = this.props;

        this.title = title;
        this.description = description;
        this.mediaId = mediaId;
    }

    @action handleMediaClick = () => {
        this.mediaOverlayOpen = true;
    };

    @action handleMediaConfirm = (media: Media) => {
        this.mediaId = media.id;
        this.mediaOverlayOpen = false;
    };

    @action handleMediaOverlayClose = () => {
        this.mediaOverlayOpen = false;
    };

    @action handleTitleChange = (title: ?string) => {
        this.title = title;
    };

    @action handleDescriptionChange = (description: ?string) => {
        this.description = description;
    };

    handleCancel = () => {
        const {id, onCancel, type} = this.props;

        onCancel(type, id);
    };

    handleReset = () => {
        const {id, onApply, type} = this.props;

        onApply({id, type});
    };

    handleApply = () => {
        const {id, onApply, type} = this.props;

        onApply({description: this.description, id, mediaId: this.mediaId, title: this.title, type});
    };

    render() {
        const {edited, editing, locale, type} = this.props;
        const {mediaUrl} = Item;

        return (
            editing
                ? <Fragment>
                    <div className={itemStyles.editForm}>
                        <div className={itemStyles.form}>
                            <div className={itemStyles.mediaColumn}>
                                {mediaUrl &&
                                    <button className={itemStyles.mediaButton} onClick={this.handleMediaClick}>
                                        {this.mediaId
                                            ? <img src={mediaUrl.replace(':id', this.mediaId.toString())} />
                                            : <MimeTypeIndicator iconSize={16} mimeType="image" />
                                        }
                                        <Icon className={itemStyles.mediaButtonIcon} name="su-pen" />
                                    </button>
                                }
                            </div>
                            <div className={itemStyles.formColumn}>
                                <div className={itemStyles.titleInput}>
                                    <Input onChange={this.handleTitleChange} value={this.title} />
                                </div>
                                <div className={itemStyles.descriptionTextArea}>
                                    <TextEditor
                                        adapter="ckeditor5"
                                        locale={locale}
                                        onChange={this.handleDescriptionChange}
                                        value={this.description}
                                    />
                                </div>
                            </div>
                        </div>
                        <div className={itemStyles.buttons}>
                            <Button onClick={this.handleCancel}>{translate('sulu_admin.cancel')}</Button>
                            <Button onClick={this.handleReset} skin="link">{translate('sulu_admin.reset')}</Button>
                            <Button onClick={this.handleApply} skin="primary">{translate('sulu_admin.apply')}</Button>
                        </div>
                    </div>
                    <SingleMediaSelectionOverlay
                        locale={locale}
                        onClose={this.handleMediaOverlayClose}
                        onConfirm={this.handleMediaConfirm}
                        open={this.mediaOverlayOpen}
                    />
                </Fragment>
                : <div className={itemStyles.item}>
                    <div className={itemStyles.media}>
                        {mediaUrl && this.mediaId && <img src={mediaUrl.replace(':id', this.mediaId.toString())} />}
                    </div>
                    <div className={itemStyles.content}>
                        <p className={itemStyles.title}>{this.title}</p>
                        <p className={itemStyles.description}>
                            {this.description && textVersion(this.description)}
                        </p>
                    </div>
                    <p className={itemStyles.type}>
                        {teaserProviderRegistry.get(type).title}
                        {edited && ' (' + translate('sulu_page.edited') + ')'}
                    </p>
                </div>
        );
    }
}

export default Item;
