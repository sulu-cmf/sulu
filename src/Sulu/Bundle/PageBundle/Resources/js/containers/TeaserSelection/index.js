// @flow
import TeaserSelection from './TeaserSelection';
import teaserProviderRegistry from './registries/teaserProviderRegistry';

export default TeaserSelection;
export {teaserProviderRegistry};
