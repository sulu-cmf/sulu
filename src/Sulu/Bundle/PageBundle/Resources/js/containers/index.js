// @flow
import TeaserSelection from './TeaserSelection';

export {TeaserSelection};
