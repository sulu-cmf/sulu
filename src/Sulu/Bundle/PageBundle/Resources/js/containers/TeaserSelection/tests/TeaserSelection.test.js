// @flow
import React from 'react';
import {observable} from 'mobx';
import {mount} from 'enzyme';
import {MultiListOverlay, TextEditor} from 'sulu-admin-bundle/containers';
import TeaserSelection from '../TeaserSelection';
import TeaserStore from '../stores/TeaserStore';
import Item from '../Item';

jest.mock('sulu-media-bundle/containers/SingleMediaSelectionOverlay', () => jest.fn(() => null));

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('sulu-admin-bundle/containers/MultiListOverlay', () => jest.fn(() => null));

jest.mock('sulu-admin-bundle/containers/TextEditor', () => jest.fn(
    ({value}) => (<textarea onChange={jest.fn()} value={value} />))
);

jest.mock('../stores/TeaserStore', () => jest.fn());

jest.mock('../registries/teaserProviderRegistry', () => ({
    keys: ['pages', 'articles'],
    get: jest.fn((key) => {
        switch (key) {
            case 'pages':
                return {title: 'Pages'};
            case 'articles':
                return {title: 'Articles'};
            case 'contacts':
                return {title: 'Contacts'};
        }
    }),
}));

beforeEach(() => {
    TeaserSelection.Item.mediaUrl = '/admin/media/:id?format=sulu-25x25';
});

test('Render loading teaser selection', () => {
    const value = {
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
        ],
        presentAs: '',
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
        this.loading = true;
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={value} />);

    teaserSelection.update();
    expect(teaserSelection.render()).toMatchSnapshot();
});

test('Render teaser selection with presentations', () => {
    const value = {
        presentAs: 'test-2',
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
        ],
    };

    const presentations = [
        {
            label: 'Test 1',
            value: 'test-1',
        },
        {
            label: 'Test 2',
            value: 'test-2',
        },
    ];

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection
            locale={observable.box('en')}
            onChange={jest.fn()}
            presentations={presentations}
            value={value}
        />
    );

    teaserSelection.update();
    expect(teaserSelection.render()).toMatchSnapshot();
});

test('Render teaser selection with data', () => {
    const value = {
        presentAs: '',
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={value} />);
    teaserSelection.instance().teaserStore.loading = false;

    teaserSelection.update();
    expect(teaserSelection.render()).toMatchSnapshot();
});

test('Render MultiItemSelection disabled when disabled flag is set', () => {
    const teaserSelection = mount(
        <TeaserSelection disabled={true} locale={observable.box('en')} onChange={jest.fn()} />
    );

    expect(teaserSelection.find('MultiItemSelection').prop('disabled')).toEqual(true);
});

test('Avoid that MultiListOverlay loads the preSelectedItems from start', () => {
    const teaserSelection = mount(
        <TeaserSelection disabled={true} locale={observable.box('en')} onChange={jest.fn()} />
    );

    expect(teaserSelection.find(MultiListOverlay)).toHaveLength(2);
    expect(teaserSelection.find(MultiListOverlay).at(0).prop('preloadSelectedItems')).toEqual(false);
    expect(teaserSelection.find(MultiListOverlay).at(1).prop('preloadSelectedItems')).toEqual(false);
});

test('Call onChange when presentation is changed', () => {
    const changeSpy = jest.fn();

    const presentations = [
        {
            label: 'Test 1',
            value: 'test-1',
        },
        {
            label: 'Test 2',
            value: 'test-2',
        },
    ];

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection
            locale={observable.box('en')}
            onChange={changeSpy}
            presentations={presentations}
            value={undefined}
        />
    );

    teaserSelection.find('Button[icon="su-eye"]').simulate('click');
    teaserSelection.find('Action[value="test-2"]').simulate('click');

    expect(changeSpy).toBeCalledWith({
        presentAs: 'test-2',
        items: [],
    });
});

test('Add passed data to TeaserStore', () => {
    const value = {
        presentAs: '',
        items: [
            {
                description: 'Description 1',
                id: 2,
                title: 'Title 1',
                type: 'pages',
            },
            {
                description: 'Description 2',
                id: 3,
                title: 'Title 2',
                type: 'contacts',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={value} />);

    expect(teaserSelection.instance().teaserStore.add).toBeCalledTimes(2);
    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 2);
    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('contacts', 3);
});

test('Load combined data from TeaserStore and props', () => {
    const value = {
        presentAs: '',
        items: [
            {
                description: 'Edited Page Description',
                id: 2,
                title: 'Edited Page Title',
                type: 'pages',
            },
            {
                description: undefined,
                id: 3,
                title: undefined,
                type: 'contacts',
            },
            {
                id: 4,
                type: 'contacts',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();

        this.findById = jest.fn((type, id) => {
            if (type === 'pages' && id === 2) {
                return {
                    description: 'Page Description',
                    id: 2,
                    mediaId: 8,
                    title: 'Page',
                    type: 'pages',
                };
            }

            if (type === 'contacts' && id === 3) {
                return {
                    description: 'Contact Description 1',
                    id: 3,
                    title: 'Contact 1',
                    type: 'contacts',
                };
            }

            if (type === 'contacts' && id === 4) {
                return {
                    description: 'Contact Description 2',
                    id: 4,
                    title: 'Contact 2',
                    type: 'contacts',
                };
            }

            throw new Error('This case should not happen!');
        });
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={value} />);

    teaserSelection.update();
    expect(teaserSelection.render()).toMatchSnapshot();
});

test('Opening different adding overlays and close them without any action', () => {
    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={undefined} />
    );

    expect(teaserSelection.find('MultiItemSelection').prop('leftButton').options).toEqual([
        {label: 'Pages', value: 'pages'},
        {label: 'Articles', value: 'articles'},
    ]);

    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(false);
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('open')).toEqual(false);

    teaserSelection.find('MultiItemSelection').prop('leftButton').onClick('articles');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(false);
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('open')).toEqual(true);

    teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('onClose')();
    teaserSelection.find('MultiItemSelection').prop('leftButton').onClick('pages');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(true);
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('open')).toEqual(false);
});

test('Adding a teaser element', () => {
    const changeSpy = jest.fn();

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={undefined} />
    );

    teaserSelection.find('Button[icon="su-plus-circle"]').simulate('click');
    teaserSelection.find('Action[value="pages"]').simulate('click');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(true);
    teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('onConfirm')([{id: 6}, {id: 5}]);

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(false);

    expect(changeSpy).toBeCalledWith({
        presentAs: undefined,
        items: [{id: 6, type: 'pages'}, {id: 5, type: 'pages'}],
    });

    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 6);
    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 5);
});

test('Adding two different kind of teasers', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
            {id: 8, type: 'pages'},
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />
    );

    teaserSelection.find('Button[icon="su-plus-circle"]').simulate('click');
    teaserSelection.find('Action[value="articles"]').simulate('click');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('open')).toEqual(true);
    teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('onConfirm')([{id: 6}]);

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="articles"]').prop('open')).toEqual(false);

    expect(changeSpy).toBeCalledWith({
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
            {id: 8, type: 'pages'},
            {id: 6, type: 'articles'},
        ],
    });

    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('articles', 6);
});

test('Adding a teaser item along with other teaser items which has already been added', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />
    );

    teaserSelection.find('Button[icon="su-plus-circle"]').simulate('click');
    teaserSelection.find('Action[value="pages"]').simulate('click');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(true);
    teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('onConfirm')([{id: 5}, {id: 6}]);

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(false);

    expect(changeSpy).toBeCalledWith({
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
            {id: 6, type: 'pages'},
        ],
    });

    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 6);
});

test('Removing by unselecting element in teaser selection', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
            {id: 8, type: 'pages'},
            {id: 5, type: 'articles'},
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />
    );

    teaserSelection.find('Button[icon="su-plus-circle"]').simulate('click');
    teaserSelection.find('Action[value="pages"]').simulate('click');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(true);
    teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('onConfirm')([{id: 6}]);

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(false);

    expect(changeSpy).toBeCalledWith({
        presentAs: undefined,
        items: [{id: 5, type: 'articles'}, {id: 6, type: 'pages'}],
    });

    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 6);
    expect(teaserSelection.instance().teaserStore.add).toBeCalledWith('pages', 5);
});

test('Preselecting correct items', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: undefined,
        items: [
            {id: 5, type: 'pages'},
            {id: 8, type: 'pages'},
            {id: 5, type: 'articles'},
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(
        <TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />
    );

    teaserSelection.find('Button[icon="su-plus-circle"]').simulate('click');
    teaserSelection.find('Action[value="pages"]').simulate('click');

    teaserSelection.update();
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('open')).toEqual(true);
    expect(teaserSelection.find(MultiListOverlay).find('[resourceKey="pages"]').prop('preSelectedItems'))
        .toEqual([{id: 5, type: 'pages'}, {id: 8, type: 'pages'}]);
});

test('Open and close items when clicking on the pen icon', () => {
    const value = {
        presentAs: '',
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
            {
                description: 'Description 2',
                id: 6,
                title: 'Title 2',
                type: 'pages',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={jest.fn()} value={value} />);

    expect(teaserSelection.find(Item).at(0).prop('editing')).toEqual(false);
    expect(teaserSelection.find(Item).at(1).prop('editing')).toEqual(false);

    teaserSelection.find('Icon[name="su-pen"]').at(0).parent().prop('onClick')();
    teaserSelection.update();

    expect(teaserSelection.find(Item).at(0).prop('editing')).toEqual(true);
    expect(teaserSelection.find(Item).at(1).prop('editing')).toEqual(false);

    teaserSelection.find('Icon[name="su-pen"]').at(1).parent().prop('onClick')();
    teaserSelection.update();

    expect(teaserSelection.find(Item).at(0).prop('editing')).toEqual(true);
    expect(teaserSelection.find(Item).at(1).prop('editing')).toEqual(true);

    teaserSelection.find('Button[children="sulu_admin.cancel"]').at(0).prop('onClick')();
    teaserSelection.update();

    expect(teaserSelection.find(Item).at(0).prop('editing')).toEqual(false);
    expect(teaserSelection.find(Item).at(1).prop('editing')).toEqual(true);
});

test('Call onChange with new values when apply button is clicked', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: '',
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
            {
                description: 'Description 2',
                id: 6,
                title: 'Title 2',
                type: 'pages',
            },
            {
                description: 'Description 3',
                id: 6,
                title: 'Title 3',
                type: 'contacts',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />);

    teaserSelection.find('Icon[name="su-pen"]').at(1).parent().prop('onClick')();
    teaserSelection.update();

    teaserSelection.find('Input').prop('onChange')('Edited Title 2');
    teaserSelection.find(TextEditor).prop('onChange')('Edited Description 2');

    teaserSelection.find('Button[children="sulu_admin.apply"]').prop('onClick')();

    expect(changeSpy).toBeCalledWith(
        {
            presentAs: '',
            items: [
                {
                    description: 'Description',
                    id: 2,
                    title: 'Title',
                    type: 'pages',
                },
                {
                    description: 'Edited Description 2',
                    id: 6,
                    title: 'Edited Title 2',
                    type: 'pages',
                },
                {
                    description: 'Description 3',
                    id: 6,
                    title: 'Title 3',
                    type: 'contacts',
                },
            ],
        }
    );
});

test('Call onChange with new values after one item is removed', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: '',
        items: [
            {
                description: 'Contact',
                id: 6,
                title: 'Contact',
                type: 'contacts',
            },
            {
                description: 'Description 2',
                id: 6,
                title: 'Title 2',
                type: 'pages',
            },
            {
                description: 'Description 3',
                id: 7,
                title: 'Title 3',
                type: 'pages',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />);

    teaserSelection.find('Icon[name="su-trash-alt"]').at(1).parent().prop('onClick')();

    expect(changeSpy).toBeCalledWith(
        {
            presentAs: '',
            items: [
                {
                    description: 'Contact',
                    id: 6,
                    title: 'Contact',
                    type: 'contacts',
                },
                {
                    description: 'Description 3',
                    id: 7,
                    title: 'Title 3',
                    type: 'pages',
                },
            ],
        }
    );
});

test('Call onChange with new values after items are sorted', () => {
    const changeSpy = jest.fn();

    const value = {
        presentAs: '',
        items: [
            {
                description: 'Description',
                id: 2,
                title: 'Title',
                type: 'pages',
            },
            {
                description: 'Description 2',
                id: 6,
                title: 'Title 2',
                type: 'pages',
            },
            {
                description: 'Description 3',
                id: 9,
                title: 'Title 3',
                type: 'pages',
            },
        ],
    };

    // $FlowFixMe
    TeaserStore.mockImplementation(function() {
        this.add = jest.fn();
        this.findById = jest.fn();
    });

    const teaserSelection = mount(<TeaserSelection locale={observable.box('en')} onChange={changeSpy} value={value} />);

    teaserSelection.find('MultiItemSelection').prop('onItemsSorted')(2, 1);

    expect(changeSpy).toBeCalledWith(
        {
            presentAs: '',
            items: [
                {
                    description: 'Description',
                    id: 2,
                    title: 'Title',
                    type: 'pages',
                },
                {
                    description: 'Description 3',
                    id: 9,
                    title: 'Title 3',
                    type: 'pages',
                },
                {
                    description: 'Description 2',
                    id: 6,
                    title: 'Title 2',
                    type: 'pages',
                },
            ],
        }
    );
});
