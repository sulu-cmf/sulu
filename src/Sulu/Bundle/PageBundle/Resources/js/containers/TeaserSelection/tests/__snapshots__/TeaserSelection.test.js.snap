// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load combined data from TeaserStore and props 1`] = `
<div
  class="multiItemSelectionClass"
>
  <div
    class="header"
  >
    <button
      class="button left hasOptions"
      type="button"
    >
      <span
        aria-label="su-plus-circle"
        class="su-plus-circle icon"
      />
      <span
        aria-label="su-angle-down"
        class="su-angle-down"
      />
    </button>
    <div
      class="label"
    />
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            >
              <img
                src="/admin/media/8?format=sulu-25x25"
              />
            </div>
            <div
              class="content"
            >
              <p
                class="title"
              >
                Edited Page Title
              </p>
              <p
                class="description"
              >
                Edited Page Description

              </p>
            </div>
            <p
              class="type"
            >
              Pages (sulu_page.edited)
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            2
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            />
            <div
              class="content"
            >
              <p
                class="title"
              >
                Contact 1
              </p>
              <p
                class="description"
              >
                Contact Description 1

              </p>
            </div>
            <p
              class="type"
            >
              Contacts
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            3
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            />
            <div
              class="content"
            >
              <p
                class="title"
              >
                Contact 2
              </p>
              <p
                class="description"
              >
                Contact Description 2

              </p>
            </div>
            <p
              class="type"
            >
              Contacts
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Render loading teaser selection 1`] = `
<div
  class="multiItemSelectionClass"
>
  <div
    class="header"
  >
    <button
      class="button left hasOptions"
      type="button"
    >
      <span
        aria-label="su-plus-circle"
        class="su-plus-circle icon"
      />
      <span
        aria-label="su-angle-down"
        class="su-angle-down"
      />
    </button>
    <div
      class="label"
    >
      <div
        class="loader"
      >
        <div
          class="spinner"
          style="width: 24px; height: 24px;"
        >
          <div
            class="doubleBounce1"
          />
          <div
            class="doubleBounce2"
          />
        </div>
      </div>
    </div>
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            />
            <div
              class="content"
            >
              <p
                class="title"
              >
                Title
              </p>
              <p
                class="description"
              >
                Description

              </p>
            </div>
            <p
              class="type"
            >
              Pages (sulu_page.edited)
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Render teaser selection with data 1`] = `
<div
  class="multiItemSelectionClass"
>
  <div
    class="header"
  >
    <button
      class="button left hasOptions"
      type="button"
    >
      <span
        aria-label="su-plus-circle"
        class="su-plus-circle icon"
      />
      <span
        aria-label="su-angle-down"
        class="su-angle-down"
      />
    </button>
    <div
      class="label"
    />
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            />
            <div
              class="content"
            >
              <p
                class="title"
              >
                Title
              </p>
              <p
                class="description"
              >
                Description

              </p>
            </div>
            <p
              class="type"
            >
              Pages (sulu_page.edited)
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Render teaser selection with presentations 1`] = `
<div
  class="multiItemSelectionClass"
>
  <div
    class="header"
  >
    <button
      class="button left hasOptions"
      type="button"
    >
      <span
        aria-label="su-plus-circle"
        class="su-plus-circle icon"
      />
      <span
        aria-label="su-angle-down"
        class="su-angle-down"
      />
    </button>
    <div
      class="label"
    />
    <button
      class="button right hasLabel hasOptions"
      type="button"
    >
      <span
        aria-label="su-eye"
        class="su-eye icon"
      />
      <span
        class="label"
      >
        Test 2
      </span>
      <span
        aria-label="su-angle-down"
        class="su-angle-down"
      />
    </button>
  </div>
  <ul
    class="list"
  >
    <li
      class="listElement"
    >
      <div
        class="item"
      >
        <span
          class="dragHandle sortable"
        >
          <span
            aria-label="su-more"
            class="su-more"
          />
          <span
            class="index"
          >
            1
          </span>
        </span>
        <div
          class="content"
        >
          <div
            class="item"
          >
            <div
              class="media"
            />
            <div
              class="content"
            >
              <p
                class="title"
              >
                Title
              </p>
              <p
                class="description"
              >
                Description

              </p>
            </div>
            <p
              class="type"
            >
              Pages (sulu_page.edited)
            </p>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-pen"
              class="su-pen"
            />
          </button>
          <button
            class="button"
            type="button"
          >
            <span
              aria-label="su-trash-alt"
              class="su-trash-alt"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
